src/main/c/NativeModule.cpp

```diff
@@ -0,0 +1,713 @@
+    ModuleBuilder builder(this, options, source, include, key_);
+    builder.try_to_start_build();
+  }
+}
+
+NativeModule::NativeModule(
+  bool is_global,
+  const char* key,
+  long binary_size,
+  const void* binary
+) :
+  build_state_(is_global ? kPass : kInit),
+  load_state_(is_global ? kPass : kInit),
+  key_(key),
+  is_global_(is_global),
+  dlopen_handle_(nullptr),
+  lock_name_(config.get_lock_name(key_)),
+  lib_name_(config.get_lib_name(key_)),
+  new_name_(config.get_new_name(key_)) {
+  // Worker constructor - try to get the binary written to local file
+  if (is_global_) return;
+  int rc = 0;
+  config.ensure_module_dir_exists();
+  auto mylock = std::make_shared< std::lock_guard<NativeModule> >(*this);
+  for (;;) {
+    struct stat lib_stat;
+    if (file_stat(lib_name_, &lib_stat) && (lib_stat.st_size == binary_size)) {
+      build_state_ = kPass;
+      break;
+    }
+    // Race to write the new file
+    int fd = open(new_name_.c_str(), O_WRONLY|O_CREAT|O_EXCL, 0666);
+    if (fd >= 0) {
+      // Now we're about to write the new file
+      rc = write(fd, binary, binary_size);
+      assert(rc == binary_size);
+      ::close(fd);
+      ::chmod(new_name_.c_str(), 0644);
+      struct stat st;
+      if (file_stat(lib_name_, &st)) {
+        long old_size = st.st_size;
+        if (old_size == binary_size) {
+          ::unlink(new_name_.c_str());
+          break;
+        } else if (old_size == 0) {
+          ::unlink(lib_name_.c_str());
+        } else {
+          auto old = lib_name_ + ".old";
+          ::rename(lib_name_.c_str(), old.c_str());
+        }
+      }
+      // Don't let anyone see the file until it is completely written
+      rc = ::rename(new_name_.c_str(), lib_name_.c_str());
+      build_state_ = ((rc == 0) ? kPass : kFail);
+      break;
+    } else {
+      // Someone else is writing to new
+      while (file_exists_and_is_recent(new_name_)) {
+        usleep_without_lock(kFilePollMicrosecs);
+      }
+    }
+  }
+  mylock.reset();
+  if (build_state_ == kPass) {
+    try_load();
+  }
+}
+
+NativeModule::~NativeModule() {
+  if (!is_global_ && dlopen_handle_) {
+    dlclose(dlopen_handle_);
+  }
+}
+
+// We may have many threads across many processes trying to access the shared
+// files in ~/hail_modules.  We protect the files for each module by
+// mutual exclusion based on the existence of a link hm_${key}.lock
+//
+// To achieve disaster recovery after a killed process or crash, a .lock
+// link will be ignored if it is more than 20 seconds old.  The critical
+// sections protected by the lock should be much shorter than this.
+//
+// Permission to build a new file is controlled by the existence of hm_${key}.new,
+// and has a longer timeout (120 sec) to allow for module compile time.
+//
+// The intent is that when hail is not running, it should be ok to do
+// "rm ~/hail_modules/*" to clear the cache and start again.
+
+void NativeModule::lock() {
```

Comment:
Is there a reason to not use the `flock` system call on a file named `lock_name_` where competing threads can race to create `lock_name_` if it doesn't exist?

Reply:
So glad you asked that question.  I was feeling especially paranoid after finding out that
MacOS "/bin/mv -f" doesn't use an atomic rename(), and so has a window when the old file has
disappeeared but the new one hasn't replaced it.  So I didn't want to venture into the minefield
of Unix file locking.

Here's a summary of how much of a minefield it is: http://0pointer.de/blog/projects/locking.html

"Here's a list of places where you shouldn't use file locking due to the problems shown above: If you want to lock a file in $HOME, forget about it as $HOME might be NFS and locks generally are not reliable there. The same applies to every other file system that might be shared across the network. If the file you want to lock is accessible to more than your own user (i.e. an access mode > 0700), forget about locking, it would allow others to block your application indefinitely. If your program is non-trivial or threaded or uses a framework such as Gtk+ or Qt or any of the module-based APIs such as NSS, PAM, ... forget about about POSIX locking. If you care about portability, don't use file locking.
...
Or in one sentence: in its current state Linux file locking is unusable."

Reply:
Some resources on this issue that I drew upon:

 - "On the Brokenness of File Locking", http://0pointer.de/blog/projects/locking.html
 - "Addendum on the Brokenness of File Locking", http://0pointer.de/blog/projects/locking2
 - "File locking in Linux", https://gavv.github.io/blog/file-locks/
 - "GLIBC Manual 13.16 File Locks", https://www.gnu.org/software/libc/manual/html_node/File-Locks.html

There's a number of issues unique to `fcntl` and `lockf` that we can completely ignore because we do not need record- or byte-range locking on files. We need to lock the entirety of a single file.

Let's focus on BSD-style locking, `flock`.

 - the lock is tied to the "file object", basically to a single file description, unless the file descriptor is duplicated with `dup2` or `fork`

 - malicious or buggy co-tenants could DOS our access to the file lock, if the permissions are too permissive

 - basically broken on NFS volumes. totally broken on NFS volumes on GNU/Linux before 2.6.11 (released 2006 ish, 2.4.x EOL'ed December of 2011) and mostly broken after this because it is emulated using `fcntl` which ties locks to a pid rather than a fd.

 - no atomic mode-switch from exclusive to shared

---

Pid Eins (0pointer.de) spends a lot of time hammering NFS file locking, but this seems irrelevant to us. We are not trying to lock well known files, like `/etc/passwd`, we are using private locks for our own purposes so we can avoid both the permissions issues and the NFS issues by simply placing files in `/tmp` and setting permissions to 600. Moreover we don't need atomic mode-switch.

The core issue here seems to be lock implementations on NFS file systems behaving badly. I don't think any deployment environment that we care about will mount an NFS share at `/tmp`. Moreover, even if we were to store these in `/home`, google cloud systems do not use NFS for `/home`, and I don't think Amazon does either.

Reply:
@rcownie wrote:

> The default is to put the modules in ${HOME}/hail_modules, which in many
environments would indeed be an NFS directory.  It only goes to /tmp/hail_modules if
${HOME} is undefined.
>
> The thinking behind this is that there's a huge amount of re-use of code
for an individual from one Hail analysis to the next, but probably much less overlap between different users. And while multi-user sharing ought to work, it raises potential issues about file access permissions which seemed like trouble without a clear benefit.

The distinction between multiple users is not clear to me within Spark. AFAIK, one OS user will own the Spark leader process which orchestrates jobs for a variety of Spark users.

---

The core issue here seems to be concern that the hail_modules directory will be an NFS drive and that on an NFS drive `flock` acts like `fcntl` (I think we can safely reject anyone running an EOL'ed kernel from 2011) meaning that if we have multiple threads they cannot lock files separately, locks are associated with PIDs.

If we have multiple compilations happening in the same process, we could potentially have concurrent builds running.

It seems like we can side-step custom lock code and the `fcntl`/pid issue by ensuring we use a non-NFS drive, which is trivial in cloud Spark deployments because neither `/home` nor `/tmp` will be an NFS drive and when we move to k8s deployments the same is true. It's possible that on-prem Spark clusters will mount an NFS drive for the `home` directory of the service user that owns the Spark executor process (I believe this is the case for Broad's on-prem Spark cluster). In such a deployment, we'd have to use `/tmp` or `/var` to be resilient to multiple threads trying to compile at once.

---

Perhaps it is also worth noting that in these deployments, there will be one leader node doing all compilation, shared among all users.

---

I feel like there's a lot of conversation going on over a pretty small piece of code. It mostly does not matter to me, but I think our medium-term plan is to move to an environment where NFS shares are completely irrelevant (because every Hail process will run in a container with it's own private file system). Couple that with my strong desire to use existing well-tested software as much as possible, and that's sort of why I find `flock` compelling.

Reply:
>The distinction between multiple users is not clear to me within Spark. AFAIK, one OS user will >own the Spark leader process which orchestrates jobs for a variety of Spark users.

Well, when I run tests I see the files showing up in ~rcownie/hail_modules.  Which makes me
think that in at least some of our environments, two users who happened to be sharing a machine
would each see their files showing up in their own $HOME/hail_modules.  Supporting that isn't a major goal - but it seems highly likely that we need it to work smoothly, even if only for possible
configurations of our CI test infrastructure.

Now no-one has so far spelled out the benefit of of using some explicit file-lock mechanism -
and to my mind there's only one real benefit, which is to have locks which go away when the
lock-holding process dies, without having somewhat arbitrary timeouts.  That would be nice.  Against that is the difficulty of finding a file-locking mechanism which actually works across both 
MacOS and various Linux'es, and across various filesystem environments, with the further 
restriction in the current design that you need to be able to lock/unlock from a command in a 
Makefile.  That aspect could be designed away, at the cost of having more complicated state 
machines and process-spawning in the NativeModule.cpp,  but then is it a win ?  I'm not sure.

Another option is to say "we shouldn't be doing this, just use the open-source ccache which
already handles it".  And I'd be ok with that except that it then adds another installation/packaging
requirement, and we don't have a particularly smooth way of handling package dependencies
and installation right now as far as I can see.  And I don't know for sure whether ccache works
with MacOS, so that would be a little work to find out whether it's even an option.

So there are a whole lot of possible alternative designs - as always - but all the ones I've
thought about involve some extra complexity somewhere.  Because file locking on Linux really
is a mess.  And I'm suspicious that file locking on MacOS might actually be even *more* of a
mess, based on stuff I've read about HFS+ which didn't even have an atomic rename for a long time.  It never got much love as a server OS.